org.eclipse.jdt.ls.core/gradle/android/init.gradle

```diff
@@ -0,0 +1,430 @@
+
+        AddSourceFoldersAction(Project project) {
+            this.project = project
+        }
+
+        @Override
+        void execute(Classpath classpath) {
+            for (String key : SUPPORTED_SOURCE_SETS.keySet()) {
+                Object sourceSet = getAndroidSourceSetByName(project, key)
+                if (sourceSet == null) {
+                    continue
+                }
+                if (sourceSet.properties.containsKey("javaDirectories")) {
+                    Object javaDirectories = sourceSet.properties.get("javaDirectories")
+                    if (javaDirectories instanceof Collection) {
+                        for (Object javaDirectory : (Collection) javaDirectories) {
+                            if (javaDirectory instanceof File) {
+                                addFolderToSourceFolder(classpath, project, (File) javaDirectory, SUPPORTED_SOURCE_SETS.get(key))
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private static class AddDataBindingFilesAction implements Action<Classpath> {
+        private final Project project
+        private final FileReferenceFactory fileReferenceFactory
+
+        AddDataBindingFilesAction(Project project) {
+            this.project = project
+            this.fileReferenceFactory = new FileReferenceFactory()
+        }
+
+        @Override
+        void execute(Classpath classpath) {
+            for (Object variant : getAndroidDebuggableVariants(project)) {
+                try {
+                    // TODO: get dataBinding task from variants?
+                    String variantName = variant.properties.get("name")
+                    String taskName = "dataBindingGenBaseClasses" + variantName.substring(0, 1).toUpperCase() + variantName.substring(1)
+                    Task dataBindingGenTask = project.tasks.getByName(taskName)
+                    Object outFolder = dataBindingGenTask.properties.get("sourceOutFolder").properties.get("orNull").properties.get("asFile")
+                    if (outFolder instanceof File) {
+                        addFolderToSourceFolder(classpath, project, (File) outFolder, DEFAULT_OUTPUT_MAIN)
+                    }
+                } catch (NullPointerException | UnknownTaskException ignored) {
+                    // NPE occurs when the variant doesn't have related properties, we just skip these unsupported scenarios
+                }
+            }
+        }
+    }
+
+    private static class AddRFilesAction implements Action<Classpath> {
+        private final Project project
+        private final FileReferenceFactory fileReferenceFactory
+
+        AddRFilesAction(Project project) {
+            this.project = project
+            this.fileReferenceFactory = new FileReferenceFactory()
+        }
+
+        @Override
+        void execute(Classpath classpath) {
+            for (Object variant : getAndroidDebuggableVariants(project)) {
```

Comment:
I noticed `getAndroidDebuggableVariants(project)` being called multiple times in difference `Action`s. Can we just fetch the variants once and pass it as an argument to all the constructors of the action classes?

Reply:
Sure. I'll change it.